src/relations/mysql.py

```diff
@@ -81,8 +105,38 @@ def _on_leader_elected(self, _) -> None:
             primary_address = self.charm._mysql.get_cluster_primary_address().split(":")[0]
             relation_databag[self.charm.unit]["host"] = primary_address
 
+    def _on_config_changed(self, _) -> None:
+        """Handle the change of the username/database in config."""
+        if not self.charm.unit.is_leader():
+            return
+
+        config_username = self.charm.config.get(MYSQL_RELATION_USER_KEY)
+        config_database = self.charm.config.get(MYSQL_RELATION_DATABASE_KEY)
+        if not config_username and not config_database:
+            return
+
+        current_username = self.charm.app_peer_data.get(MYSQL_RELATION_USER_KEY)
+        current_database = self.charm.app_peer_data.get(MYSQL_RELATION_DATABASE_KEY)
+
+        # the mysql-relation-created event has not yet fired
+        # the config options will be factored in when the relation is formed
+        if not current_username:
+            return
+
+        mysql_relation_exists = len(self.model.relations[LEGACY_MYSQL])
+
+        if isinstance(self.charm.unit.status, ActiveStatus) and mysql_relation_exists:
+            if config_username != current_username:
+                self.charm.unit.status = BlockedStatus(
+                    f"Remove `mysql` relations in order to change `{MYSQL_RELATION_USER_KEY}` config"
+                )
+            if config_database != current_database:
+                self.charm.unit.status = BlockedStatus(
+                    f"Remove `mysql` relations in order to change `{MYSQL_RELATION_DATABASE_KEY}` config"
+                )
```

Comment:
somewhat addressed in a2da4f9c678103153b48ff7a7a611e0f24e12fba. config-changed event handler should really not be setting default value in the app_peer_data. additionally, there should be checks for when the app is set to `BlockedStatus`

Reply:
> Does setting the value in app_peer_data have side effects for other units?
I do not think it has side effects. However, I am not sure it is the correct behavior when `config_value` is empty (the value in app_peer_data will not be reset). I think it is simpler to trace the way it is right now (`config-changed` just sets app status to BlockedStatus, while the username/database is determined when the relation is formed)

> Is this necessary? If they change the config value back to the old value, no need to stay in blocked imo—or is this referring to another BlockedStatus?
Some of them are necessary, e.g. only the leader should set status(?), we shouldnt override app status if it is not active, and if there are no `mysql` relations

Reply:
i understand, but if the config value is `None` (or false-y), we should be resetting the value that `setdefault` set previously. this is so that a new database/username can be autogenerated

consider the following:
1. we set the config value to `A` -> app_peer_data value will be set to A
2. we set the config value to `` -> app_peer_data value never reset

Reply:
the latter. are you implying that the app_peer_data is reset when the relation is broken (so we dont need to reset)?